packages/backend/src/open_payments/setup_payment/service.ts

```diff
@@ -0,0 +1,122 @@
+import assert from 'assert'
+import axios from 'axios'
+import { Logger } from 'pino'
+import {
+  IncomingPaymentJSON,
+  IlpStreamConnectionJSON
+} from '../../open_payments/payment/incoming/model'
+import { PaymentError } from '@interledger/pay'
+import { IlpPlugin } from '../../shared/ilp_plugin'
+import * as Pay from '@interledger/pay'
+
+import { OpenAPI, HttpMethod, ValidateFunction } from 'openapi'
+
+export interface SetupPaymentService {
+  queryIncomingPayment(
+    url: string,
+    token: string
+  ): Promise<IncomingPaymentJSON | PaymentError>
+  setupPayment(destinationPayment: string, gnapToken: string, plugin: IlpPlugin)
+}
+
+interface ServiceDependencies {
+  openApi: OpenAPI
+  logger: Logger
+  validateResponse: ValidateFunction<IncomingPaymentJSON>
+}
+
+export async function createSetupPaymentService(
+  deps_: Omit<ServiceDependencies, 'validateResponse'>
+): Promise<SetupPaymentService> {
+  const log = deps_.logger.child({
+    service: 'SetupPaymentService'
+  })
+  const validateResponse = deps_.openApi.createResponseValidator<IncomingPaymentJSON>(
+    {
+      path: '/{accountId}/incoming-payments/{incomingPaymentId}',
+      method: HttpMethod.GET
+    }
+  )
+  const deps: ServiceDependencies = {
+    ...deps_,
+    logger: log,
+    validateResponse
+  }
+
+  const createHttpUrl = (rawUrl: string, base?: string): URL | undefined => {
+    try {
+      const url = new URL(rawUrl, base)
+      if (url.protocol === 'https:' || url.protocol === 'http:') {
+        return url
+      }
+    } catch (_) {
+      return
+    }
+  }
+
+  const setupPayment = async (
+    incomingPaymentUrl: string,
+    gnapToken: string,
+    plugin: IlpPlugin
+  ): Promise<Pay.ResolvedPayment> => {
+    const incomingPayment = await queryIncomingPayment(
+      incomingPaymentUrl,
+      gnapToken
+    )
+    return await Pay.setupPayment({
```

Comment:
@DonChangfoot suggested that maybe after resolving the payment details we don't need to call Pay.setupPayment at all, since it only seems to resolve the address and secret.